Add misc macro for vector extensions. (#21)
riscv-non-isa/riscv-c-api-doc: riscv-c-api.md

| __riscv_xlen        | <ul><li>32 for rv32</li><li>64 for rv64</li><li>128 for rv128</ul> | Always defined.             |
| __riscv_flen        | <ul><li>32 if the F extension is available **or**</li><li>64 if `D` extension available **or**</li><li>128 if `Q` extension available</li></ul> | `F` extension is available. |
| __riscv_32e         | 1     | `E` extension is available.   |
| __riscv_v_min_vlen    | <N> (see [__riscv_v_min_vlen](#__riscv_v_min_vlen)) | The `V` extension or one of the `Zve*` extensions is available. |
| __riscv_v_elen     | <N> (see [__riscv_v_elen](#__riscv_v_elen)) | The `V` extension or one of the `Zve*` extensions is available. |

> | __riscv_v_elen_fp  | <N> (see [__riscv_v_elen_fp](#__riscv_v_elen_fp)) | The `V` extension or one of the `Zve*` extensions is available. |

Reviewer: The when defined for `__riscv_v_elen_fp` isn't quite right. zve32x/zve64x don't define this.

Author: `zve32x` and `zve64x` still define that but defined as `0`

Reviewer: Oops I missed that. Thanks.

| __riscv_xlen        | <ul><li>32 for rv32</li><li>64 for rv64</li><li>128 for rv128</ul> | Always defined.             |
| __riscv_flen        | <ul><li>32 if the F extension is available **or**</li><li>64 if `D` extension available **or**</li><li>128 if `Q` extension available</li></ul> | `F` extension is available. |
| __riscv_32e         | 1     | `E` extension is available.   |

> | __riscv_v_min_vlen    | <N>   | `V` or `Zve*` extensions are available, more detailed description see [__riscv_v_min_vlen](#__riscv_v_min_vlen) |
> | __riscv_v_max_eew     | <N>   | `V` or `Zve*` extensions are available, more detailed description see [__riscv_v_max_eew](#__riscv_v_max_eew) |
> | __riscv_v_max_eew_fp  | <N>   | `V` or `Zve*` extensions are available, more detailed description see [__riscv_v_max_eew_fp](#__riscv_v_max_eew_fp) |

Reviewer 2: ```suggestion
| __riscv_v_min_vlen    | <N> (see [__riscv_v_min_vlen](#__riscv_v_min_vlen)) | The `V` extension or one of the `Zve*` extension is available. |
| __riscv_v_max_eew     | <N> (see [__riscv_v_max_eew](#__riscv_v_max_eew)) | The `V` extension or one of the `Zve*` extensions is available. |
| __riscv_v_max_eew_fp  | <N> (see [__riscv_v_max_eew_fp](#__riscv_v_max_eew_fp)) | The `V` extension or one of the `Zve*` extensions is available. |
```

### __riscv_v_min_vlen

This macro present the minimal VLEN in bit, which is guaranteed by the
architecture extension, e.g. `V` extension requires that VLEN ≥ 128, `Zve32*`

> extension requires VLEN ≥ 32 and `Zve64*` extension requires VLEN ≥ 32; also

Reviewer 3: ```suggestion
extension requires VLEN ≥ 32 and `Zve64*` extension requires VLEN ≥ 64; also
```

RISC-V have additional `Zvl*b` extensions to guarantee the minimal VLEN, that
will effect the value of `__riscv_v_min_vlen` too.

> So the value of `__riscv_v_min_vlen` is defined by following rules:
> 
> - 128 if `V` extension is available.
> - 32 if `Zve32x` or `Zve32f` extension is available.
> - 64 if `Zve64x`, `Zve64f` or `Zve64d` extension is available.
> - `N` if `Zvl<N>b` extension is available, and if `N` is large than the value
>   defined above.
> - `__riscv_v_min_vlen` is undefined if all of above rule are not match.

Reviewer 2: ```suggestion
The value of `__riscv_v_min_vlen` is defined by the following rules:

- 128, if the `V` extension is present;
- 32, if one of the `Zve32{x,f}` extensions is present;
- 64, if one of the `Zve64{x,f,d}` extensions is present;
- `N`, if one of the `Zvl<N>b` extensions, `N` in `{32,64,128,256,512,1024}`, is present.

If multiple rules apply, the maximum value is taken.
If none of the rules apply, `__riscv_v_min_vlen` is undefined.
```

> Example:

Reviewer 2: ```suggestion
Examples:
```

- `__riscv_v_min_vlen` is 128 for `rv64gcv`

> - `__riscv_v_min_vlen` is 512 for `rv32gcv_zv512b`

Reviewer 2: ```suggestion
- `__riscv_v_min_vlen` is 512 for `rv32gcv_zvl512b`
```

> - `__riscv_v_min_vlen` is 256 for `rv32gcv_zvl32b_zv256b`

Reviewer 2: ```suggestion
- `__riscv_v_min_vlen` is 256 for `rv32gcv_zvl32b_zvl256b`
```

- `__riscv_v_min_vlen` is 128 for `rv64gcv_zvl32b`

### __riscv_v_max_eew

> This marco present the maximal EEW (effective element width) value for

Reviewer 2: "macro"

> This macro present the maximal EEW (effective element width) value for
> non-floating-point operations, which is guaranteed by the architecture
> extension, e.g. `V` and `Zve64*` extension require EEW of 8, 16, and 32, and 64,
> but `Zve32*` only require EEW of 8, 16, and 32.

Reviewer 2: ```suggestion
The `__riscv_v_max_eew` macro expands to the maximum EEW (effective element width), in bits, 
of any non-floating-point vector operand of any vector instruction in the available vector extension, if any.
(Stricter upper bounds may apply to particular operands of particular instructions.)
```

The value of `__riscv_v_max_eew`:

- `__riscv_v_max_eew` is 64 if `V`, `Zve64x`, `Zve64f` or `Zve64d` extension
  is available.
- `__riscv_v_max_eew` is 32 if `Zve32x` or `Zve32f` extension is available.

> - `__riscv_v_max_eew` is undefined if all of above rule are not match.

Reviewer 4: I have a question on why `__riscv_v_max_eew` is undefined here. Although it seems that no vector functions is enabled if this macro is undefined. Maybe still set it as zero for consistency with `__riscv_v_max_eew_fp`?

Author: - `__riscv_v_max_eew_fp` is undefined if all of above rule are not match.
`__riscv_v_max_eew_fp` has same rule like this, undefined if no `V` or `Zve*` enabled.

Reviewer 4: I see. Thank you for explaining.

> The value of `__riscv_v_max_eew`:
> 
> - `__riscv_v_max_eew` is 64 if `V`, `Zve64x`, `Zve64f` or `Zve64d` extension
>   is available.
> - `__riscv_v_max_eew` is 32 if `Zve32x` or `Zve32f` extension is available.
> - `__riscv_v_max_eew` is undefined if all of above rule are not match.

Reviewer 2: ```suggestion
The value of `__riscv_v_max_eew` is defined by the following rules:

- 64, if the `V` extension or one of the `Zve64{x,f,d}` extensions is present; and
- 32, if one of the `Zve32{x,f}` extensions is present.

If multiple rules apply, the maximum value is taken.
If none of the rules apply, `__riscv_v_max_eew` is undefined.
```

> NOTE: Toolchain may support integer vector type large than `__riscv_v_max_eew`,
>       but it might cause illegal instruction exception at run-time.

Reviewer 2: The technical term "vector type" is not currently defined in this document (or the C language) so I think we should avoid using/discussing it.

```suggestion
```

### __riscv_v_max_eew_fp

> This marco present the maximal EEW value for floating-point operations,

Reviewer 2: @kito-cheng My suggestion got broken somehow, I guess I must have been reviewing an older commit?

> This macro present the maximal EEW value for floating-point operations,
> which is guaranteed by the architecture extension, e.g. Support of
> floating-point operands with EEW=32 or EEW=64 are required by `V` extension and
> `zve64d` extension, `Zve32f` and `Zve64f` only require support
> floating-point operations with EEW=32, and `Zve32x` and `Zve64x` are not supprot
> floating-point operations.

Reviewer 2: ```suggestion
The `__riscv_v_max_eew_fp` macro expands to the maximum EEW (effective element width), in bits, 
of any floating-point vector operand of any vector instruction in the available vector extension, if any.
(Stricter upper bounds may apply to particular operands of particular instructions.)
```

The value of `__riscv_v_max_eew_fp`:

- `__riscv_v_max_eew_fp` is 64 if `V` or `Zve64d` extension is available.

> - `__riscv_v_max_eew_fp` is 32 if `Zve64f` or `Zve32d` extension is available.

Reviewer 3: ```suggestion
- `__riscv_v_max_eew_fp` is 32 if `Zve32f` or `Zve64f` extension is available.
```
I don't think `Zve32d` exists, and even then it implies that FP64 is supported?

Reviewer 3: I reversed to go from 32 to 64 but I just noticed the line below is still 64 to 32. Whatever's consistent.

Author: Thanks, I guess it might some thing wrong when I copy-past and modify :P

> - `__riscv_v_max_eew_fp` is 0 if `Zve64x` or `Zve32x` extension is available.

Reviewer 5: Why does it define as 0, instead of undefined?

Author: My intention is make sure this marco is always defined if vector extension is available, so we only need to check `__riscv_v_max_eew_fp` is larger than our requirement once we ensure the vector extension is available.

e.g.
```
#ifdef __riscv_v_max_eew
#error "NO vector extension supported"
#endif

#if __riscv_v_max_eew_fp > 64
...
#endif
```

Rather than
```
#ifdef __riscv_v_max_eew
#error "NO vector extension supported"
#endif

#if defined(__riscv_v_max_eew_fp) && __riscv_v_max_eew_fp >= 64
...
#endif
```

> The value of `__riscv_v_max_eew_fp`:
> 
> - `__riscv_v_max_eew_fp` is 64 if `V` or `Zve64d` extension is available.
> - `__riscv_v_max_eew_fp` is 32 if `Zve32f` or `Zve64f` extension is available.
> - `__riscv_v_max_eew_fp` is 0 if `Zve64x` or `Zve32x` extension is available.
> - `__riscv_v_max_eew_fp` is undefined if all of above rule are not match.

Reviewer 2: ```suggestion
The value of `__riscv_v_max_eew_fp` is defined by the following rules:

- 64, if one of the `V` or `Zve64d` extensions is present;
- 32, if one of the `Zve{32,64}f` extensions is present; and
- 0, if one of the `Zve{32,64}x` extensions is present.

If multiple rules apply, the maximum value is taken.
If none of the rules apply, `__riscv_v_max_eew_fp` is undefined.
```

> NOTE: Toolchain may support floating-point vector type large than
>       `__riscv_v_max_eew_fp`, but it might cause illegal instruction exception
>       at run-time.

Reviewer 2: As above, the technical term "vector type" is not currently defined in this document (or the C language) so I think we should avoid using/discussing it.

```suggestion
```
